flink-state-backends/flink-statebackend-forst/src/test/java/org/apache/flink/state/forst/ForStAsyncAndSyncCompatibilityTest.java

```diff
@@ -0,0 +1,234 @@
+        configuration.set(ForStOptions.PRIMARY_DIRECTORY, temporaryFolder.toURI().toString());
+        forStStateBackend = new ForStStateBackend().configure(configuration, null);
+
+        env = getMockEnvironment(temporaryFolder);
+
+        mailboxExecutor =
+                new MailboxExecutorImpl(
+                        new TaskMailboxImpl(), 0, StreamTaskActionExecutor.IMMEDIATE);
+    }
+
+    @Test
+    void testForStTransFromAsyncToSync() throws Exception {
+        ForStKeyedStateBackend<String> keyedBackend =
+                setUpAsyncKeyedStateBackend(Collections.emptyList());
+        MapStateDescriptor<Integer, String> descriptor =
+                new MapStateDescriptor<>(
+                        "testState", IntSerializer.INSTANCE, StringSerializer.INSTANCE);
+
+        MapState<Integer, String> asyncMapState =
+                keyedBackend.createState(1, IntSerializer.INSTANCE, descriptor);
+
+        context = aec.buildContext("testRecord", "testKey");
+        context.retain();
+        aec.setCurrentContext(context);
+        asyncMapState.asyncPut(1, "1");
+        context.release();
+        aec.drainInflightRecords(0);
+
+        RunnableFuture<SnapshotResult<KeyedStateHandle>> snapshot =
+                keyedBackend.snapshot(
+                        1L,
+                        System.currentTimeMillis(),
+                        env.getCheckpointStorageAccess()
+                                .resolveCheckpointStorageLocation(
+                                        1L, CheckpointStorageLocationReference.getDefault()),
+                        CheckpointOptions.forCheckpointWithDefaultLocation());
+
+        if (!snapshot.isDone()) {
+            snapshot.run();
+        }
+        SnapshotResult<KeyedStateHandle> snapshotResult = snapshot.get();
+        KeyedStateHandle stateHandle = snapshotResult.getJobManagerOwnedSnapshot();
+        IOUtils.closeQuietly(keyedBackend);
+        ForStSyncKeyedStateBackend<String> syncKeyedStateBackend =
+                createSyncKeyedStateBackend(
+                        forStStateBackend,
+                        env,
+                        StringSerializer.INSTANCE,
+                        Collections.singletonList(stateHandle));
+
+        try {
+            org.apache.flink.api.common.state.MapState<Integer, String> syncMapState =
+                    syncKeyedStateBackend.getOrCreateKeyedState(
+                            IntSerializer.INSTANCE,
+                            StateDescriptorUtils.transformFromV2ToV1(descriptor));
+            fail();
+
+            syncKeyedStateBackend.setCurrentKey("testKey");
+            ((InternalKvState) syncKeyedStateBackend).setCurrentNamespace(1);
+            Assert.assertEquals(syncMapState.get(1), "1");
+        } catch (Exception e) {
+            // Currently, ForStStateBackend does not support switching from Async to Sync, so this
+            // exception will be caught here
+            assertEquals(
```

Comment:
It's better to use assertThat.
```suggestion
           assertThat(e).isInstanceOf(ClassCastException.class);
            assertThat(e.getMessage())
                    .contains(
                            "org.apache.flink.runtime.state.v2.RegisteredKeyAndUserKeyValueStateBackendMetaInfo " 
                                    + "cannot be cast to class org.apache.flink.runtime.state.RegisteredKeyValueStateBackendMetaInfo");
```